### ISSUE TITLE
EmbeddedChannel does not work correctly if used with multiple threads

### ISSUE THREAD

**Issue by u6f6o** [netty/netty#7496]
### Expected behavior

I wrote some tests to check if a message distribution mechanism I wrote for blocking/expensive operations is working correctly. The basic idea is to spread inbound messages among a fixed number of blocking queues which will be consumed on different threads. More information can be found [here](https://stackoverflow.com/questions/47762213/is-it-valid-to-pass-netty-channels-to-a-queue-and-use-it-for-writes-on-a-differe). 

As the publication of the response message happens on a different thread, I cannot rely on EmbeddedChannels readOutbound method, as there is no guarantee that the message might have been processed already. In order to circumvent this problem, I wrote some util class that waits some time to receive the message: 

```java
public class EmbeddedChannels {

    public static <T> T readOutboundTimed(EmbeddedChannel channel, int waitTimeMs) {
        long waitTime = TimeUnit.MILLISECONDS.toNanos(waitTimeMs) + System.nanoTime();
        T result = channel.readOutbound();

        while (result == null && waitTime > System.nanoTime()) {
            result = channel.readOutbound();
            if (result != null) {
                return result;
            }
        }
        return null;
    }
}
```
I would expect that after some time the message would appear on the channel and that I would be able to read it from the EmbeddedChannel. 

### Actual behavior

Contrary to my assumption, no message is returned when I invoke the readOutbound() method. I also set the wait time to 1 hour in order to be certain it's not a timing issue but it did not help either. Furthermore, I also checked that the message is successfully published to the channel. 

After some doubts, I had a glimpse on the implementation of the EmbeddedChannel and realised that this class is not setup for handling such situations as it assumes that all operations occur on the same thread and also uses a non-thread-safe implementation of a Deque internally. 

I think this class (EmbeddedChannel) should be written in way that it also allows to deal with writes on different threads. 

### Workaround 

In order to see if the EmbeddedChannel would work in case every access would be guarded by a lock,  I introduced a subclass of the EmbeddedClass which simply adds synchronized to every method: 

```java 
public static class SynchronizedEmbeddedChannel extends EmbeddedChannel {

       // constructors omitted 

        @Override
        public synchronized void register() throws Exception {
            super.register();
        }

        @Override
        public synchronized ChannelMetadata metadata() {
            return super.metadata();
        }
}
```
Although this is quite ugly, it made my tests work for now. 

### Netty version

4.1.16.Final

### JVM version (e.g. `java -version`)
```
java -version
java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b16, mixed mode)
```

### OS version (e.g. `uname -a`)
```
uname -a
Darwin xxx-MacBook-Pro.local 16.7.0 Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64 x86_64
```

**Comment by u6f6o** [comment]
I am not sure if the approach, to wrap netty channels together with the received message and pass it to a queue which will be consumed by a different thread is a good approach after all. In case this approach does not work well with the netty mechanics or it is not recommend to solve this problem this way, I'd also be happy to learn about it.

**Comment by normanmaurer** [comment]
@u6f6o this is expected... `EmbeddedChannel` is NOT thread-safe by design.

**Comment by u6f6o** [comment]
@normanmaurer Thanks for the clarification! I recognised that it also does not allow to specify a separate EventExecutorGroup for specific channel handlers. Guess I have to re-think my approach.

**Comment by normanmaurer** [comment]
@u6f6o I am not sure what exactly you tried to do but maybe the `LocalChannel` / `LocalServerChannel` is more inline what you are looking for.

**Comment by u6f6o** [comment]
@normanmaurer Initially I wrote a channel handler that performed long running operations. I also assigned a separate EventExecutorGroup for this handler. As these long running operations are nicely isolated from each other, I thought about splitting them further, using a bunch of queues and a separated thread pool to execute these operations and when they finish, simply writing back the result to the channel. 

After all, I think I might have been a bit too trigger-happy with this approach and instead I'll check if the recommended approach, to assign a separate EventExecutorGroup might do the trick as well. 

Thx for mentioning LocalServerChannel and LocalChannel!
